source/adaptive_load/adaptive_load_controller_impl.cc

```diff
@@ -0,0 +1,444 @@
+        LoadScoringFunctionPlugin(metric_threshold.threshold_spec().scoring_function());
+    if (!scoring_function_or.ok()) {
+      errors.emplace_back(absl::StrCat("Failed to load ScoringFunction plugin: ",
+                                       scoring_function_or.status().message()));
+    }
+  }
+  for (const MetricSpec& metric_spec : spec.informational_metric_specs()) {
+    all_metric_specs.push_back(metric_spec);
+  }
+  for (const MetricSpec& metric_spec : all_metric_specs) {
+    if (plugin_from_name.contains(metric_spec.metrics_plugin_name())) {
+      std::vector<std::string> supported_metrics =
+          plugin_from_name[metric_spec.metrics_plugin_name()]->GetAllSupportedMetricNames();
+      if (std::find(supported_metrics.begin(), supported_metrics.end(),
+                    metric_spec.metric_name()) == supported_metrics.end()) {
+        errors.emplace_back(
+            absl::StrCat("Metric named '", metric_spec.metric_name(),
+                         "' not implemented by plugin '", metric_spec.metrics_plugin_name(),
+                         "'. Metrics implemented: ", absl::StrJoin(supported_metrics, ", "), "."));
+      }
+    } else {
+      errors.emplace_back(absl::StrCat(
+          "MetricSpec referred to nonexistent metrics_plugin_name '",
+          metric_spec.metrics_plugin_name(),
+          "'. You must declare the plugin in metrics_plugin_configs or use plugin ",
+          "'nighthawk.builtin'. Available plugins: ", absl::StrJoin(plugin_names, ", "), "."));
+    }
+  }
+  if (errors.size() > 0) {
+    return absl::InvalidArgumentError(absl::StrJoin(errors, "\n"));
+  }
+  return absl::OkStatus();
+}
+
+/**
+ * Queries the StepController for the current load specification.
+ *
+ * @param step_controller The step controller to be queried for the load.
+ * @param command_line_options The Nighthawk Service load proto, written if the step controller
+ * returned a value.
+ * @param output The adaptive load session output proto where any error code is stored.
+ *
+ * @return bool True if the step controller provided a value, false if an error occurred.
+ */
+bool GetCommandLineOptionsOrSetOutputError(
+    const StepController& step_controller,
+    nighthawk::client::CommandLineOptions& command_line_options,
+    AdaptiveLoadSessionOutput& output) {
+  absl::StatusOr<nighthawk::client::CommandLineOptions> command_line_options_or =
+      step_controller.GetCurrentCommandLineOptions();
+  if (!command_line_options_or.ok()) {
+    std::string message =
+        absl::StrCat("Error setting Nighthawk input: ", command_line_options_or.status().message());
+    output.mutable_session_status()->set_code(grpc::ABORTED);
+    output.mutable_session_status()->set_message(message);
+    ENVOY_LOG_MISC(info, message);
+    return false;
+  }
+  command_line_options = command_line_options_or.value();
+  return true;
+}
+
+} // namespace
+
+AdaptiveLoadSessionOutput PerformAdaptiveLoadSession(
```

Comment:
I realize that this API was already partially reviewed, but that was before we brought in absl::Status.  I question whether much of this would be easier to follow (and potentially easier to use) if you used StatusOr rather than having a proto that contains status fields.  Open to pushback here.

Reply:
Done, adopted StatusOr throughout, including the API.